File: modules/swagger-core/src/test/scala/XmlModelTest.scala
import javax.xml.bind.annotation._

import io.swagger.converter.ModelConverters
import io.swagger.models.ModelImpl
import io.swagger.util.Json
import models._
import org.junit.runner.RunWith
import org.scalatest.{FlatSpec, Matchers}
import org.scalatest.junit.JUnitRunner

import scala.beans.BeanProperty
import scala.collection.JavaConverters.mapAsScalaMapConverter

@RunWith(classOf[JUnitRunner])
class XmlModelTest extends FlatSpec with Matchers {
  it should "process an XML model attribute" in {
    val schemas = ModelConverters.getInstance().readAll(classOf[Monster])
    val model = schemas.get("Monster")

    model should not be (null)
    model.isInstanceOf[ModelImpl] should be(true)
    var xml = model.asInstanceOf[ModelImpl].getXml()

    xml should not be (null)
    xml.getName() should equal("monster")
    val property = model.getProperties().get("children")
    property should not be (null)
    xml = property.getXml()
    xml.getWrapped should equal(true)
    xml.getName() should be (null)
  }

  it should "not create an xml object" in {
    val schemas = ModelConverters.getInstance().readAll(classOf[Address])
    val model = schemas.get("Address")

    model should not be (null)
    model.isInstanceOf[ModelImpl] should be(true)

    val property = model.getProperties().get("streetNumber")
    var xml = property.getXml()

    xml should be(null)
  }

  it should "stay hidden per 534" in {
    val schemas = ModelConverters.getInstance().readAll(classOf[Issue534])
    schemas.get("Issue534").getProperties().size() should be(1)
  }

  it should "process a model with JAXB annotations" in {
    val schemas = ModelConverters.getInstance().readAll(classOf[ModelWithJAXBAnnotations])
    schemas.size should be (1)
    val model = schemas.get("ModelWithJAXBAnnotations")
    model should not be (null)
    model.isInstanceOf[ModelImpl] should be (true)
    var rootXml = model.asInstanceOf[ModelImpl].getXml()
    rootXml should not be (null)
    rootXml.getName() should equal("rootName")

    for ((name, property) <- model.getProperties.asScala) {
      name match {
        case "id" =>
          var xml = property.getXml
          xml should not be (null)
          xml.getName should be (null)
          xml.getAttribute should equal (true)
          xml.getWrapped should be (null)
        case "name" =>
          var xml = property.getXml
          xml should not be (null)
          xml.getName should be ("renamed")
          xml.getAttribute should be (null)
          xml.getWrapped should be (null)
        case "list" | "forcedElement" =>
          property.getXml should be (null)
        case "wrappedList" =>
          var xml = property.getXml
          xml should not be (null)
          xml.getName should be ("wrappedListItems")
          xml.getAttribute should be (null)
          xml.getWrapped should equal (true)
        case _ =>
          fail(s"""Property "${name}" was not expected""")
      }
    }
  }

  it should "deserialize a model" in {
    val yaml = """
---
type: "object"
properties:
  id:
    type: "string"
    xml:
      attribute: true
  name:
    type: "string"
    xml:
      name: "renamed"
  list:
    type: "array"
    items:
      type: "string"
  wrappedList:
    type: "array"
    xml:
      name: "wrappedListItems"
      wrapped: true
    items:
      type: "string"
  forcedElement:
    type: "array"
    items:
      type: "string"
xml:
  name: "rootName"
    """
    val model = io.swagger.util.Yaml.mapper().readValue(yaml, classOf[ModelImpl])

    val wrappedList = model.getProperties.get("wrappedList")
    wrappedList should not be (null)
    wrappedList.getXml() should not be (null)
    wrappedList.getXml().getName() should be ("wrappedListItems")
  }
}

@XmlRootElement(name = "monster")
class Monster {
  @BeanProperty
  var name: String = _

  @XmlElementWrapper()
  @XmlElement(name = "children")
  @BeanProperty
  var children: java.util.List[String] = _
}

File: modules/swagger-core/src/test/java/io/swagger/XmlModelTest.java
package io.swagger;

import static org.testng.Assert.assertEquals;
import static org.testng.Assert.assertNotNull;
import static org.testng.Assert.assertNull;
import static org.testng.Assert.assertTrue;
import static org.testng.Assert.fail;

import io.swagger.converter.ModelConverters;
import io.swagger.models.Address;
import io.swagger.models.Issue534;
import io.swagger.models.Model;
import io.swagger.models.ModelImpl;
import io.swagger.models.ModelWithJAXBAnnotations;
import io.swagger.models.Xml;
import io.swagger.models.properties.Property;

import org.testng.annotations.Test;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Map;

import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlElementWrapper;
import javax.xml.bind.annotation.XmlRootElement;

public class XmlModelTest {

    @Test(description = "it should process an XML model attribute")
    public void processXMLModelAttribute() {
        final Map<String, Model> schemas = ModelConverters.getInstance().readAll(Monster.class);
        final Model model = schemas.get("Monster");

        assertNotNull(model);
        assertTrue(model instanceof ModelImpl);
        Xml xml = ((ModelImpl) model).getXml();

        assertNotNull(xml);
        assertEquals(xml.getName(), "monster");
        final Property property = model.getProperties().get("children");
        assertNotNull(property);
        xml = property.getXml();
        assertTrue(xml.getWrapped());
        assertNull(xml.getName());
    }

    @Test(description = "it should not create an xml object")
    public void itShouldNotCreateXmlObject() {
        final Map<String, Model> schemas = ModelConverters.getInstance().readAll(Address.class);
        final Model model = schemas.get("Address");

        assertNotNull(model);
        assertTrue(model instanceof ModelImpl);

        final Property property = model.getProperties().get("streetNumber");
        final Xml xml = property.getXml();

        assertNull(xml);
    }

    @Test(description = "it should stay hidden per 534")
    public void stayHidden() {
        final Map<String, Model> schemas = ModelConverters.getInstance().readAll(Issue534.class);
        assertEquals(schemas.get("Issue534").getProperties().size(), 1);
    }

    @Test(description = "it should process a model with JAXB annotations")
    public void processModelWithJAXBAnnotations() {
        final Map<String, Model> schemas = ModelConverters.getInstance().readAll(ModelWithJAXBAnnotations.class);
        assertEquals(schemas.size(), 1);

        final Model model = schemas.get("ModelWithJAXBAnnotations");
        assertNotNull(model);
        assertTrue(model instanceof ModelImpl);

        final Xml rootXml = ((ModelImpl) model).getXml();
        assertNotNull(rootXml);
        assertEquals(rootXml.getName(), "rootName");

        for (Map.Entry<String, Property> entry : model.getProperties().entrySet()) {
            final String name = entry.getKey();
            final Property property = entry.getValue();
            if ("id".equals(name)) {
                final Xml xml = property.getXml();
                assertNotNull(xml);
                assertNull(xml.getName());
                assertTrue(xml.getAttribute());
                assertNull(xml.getWrapped());
            } else if ("name".equals(name)) {
                final Xml xml = property.getXml();
                assertNotNull(xml);
                assertEquals(xml.getName(), "renamed");
                assertNull(xml.getAttribute());
                assertNull(xml.getWrapped());
            } else if (Arrays.asList("list", "forcedElement").contains(name)) {
                assertNull(property.getXml());
            } else if ("wrappedList".equals(name)) {
                final Xml xml = property.getXml();
                assertNotNull(xml);
                assertEquals(xml.getName(), "wrappedListItems");
                assertNull(xml.getAttribute());
                assertTrue(xml.getWrapped());
            } else {
                fail(String.format("Unexpected property: %s", name));
            }
        }
    }

    @Test(description = "it should deserialize a model")
    public void deserializeModel() throws IOException {
        final String yaml = "---\n" +
                "type: \"object\"\n" +
                "properties:\n" +
                "  id:\n" +
                "    type: \"string\"\n" +
                "    xml:\n" +
                "      attribute: true\n" +
                "  name:\n" +
                "    type: \"string\"\n" +
                "    xml:\n" +
                "      name: \"renamed\"\n" +
                "  list:\n" +
                "    type: \"array\"\n" +
                "    items:\n" +
                "      type: \"string\"\n" +
                "  wrappedList:\n" +
                "    type: \"array\"\n" +
                "    xml:\n" +
                "      name: \"wrappedListItems\"\n" +
                "      wrapped: true\n" +
                "    items:\n" +
                "      type: \"string\"\n" +
                "  forcedElement:\n" +
                "    type: \"array\"\n" +
                "    items:\n" +
                "      type: \"string\"\n" +
                "xml:\n" +
                "  name: \"rootName\"";
        final ModelImpl model = io.swagger.util.Yaml.mapper().readValue(yaml, ModelImpl.class);

        final Property wrappedList = model.getProperties().get("wrappedList");
        assertNotNull(wrappedList);
        assertNotNull(wrappedList.getXml());
        assertEquals(wrappedList.getXml().getName(), "wrappedListItems");
    }

    @XmlRootElement(name = "monster")
    class Monster {
        public String name = "";

        @XmlElementWrapper()
        @XmlElement(name = "children")
        public java.util.List<String> children = new ArrayList<String>();
    }
}
